core/src/main/java/org/apache/calcite/util/format/PostgresqlDateTimeFormatter.java

```diff
@@ -0,0 +1,660 @@
+          dt -> {
+            final String yearString =
+                String.format(Locale.ROOT, "%03d", dt.get(IsoFields.WEEK_BASED_YEAR));
+            return yearString.substring(yearString.length() - 3);
+          },
+          "IYY"),
+      new NumberFormatPattern(
+          dt -> {
+            final String yearString =
+                String.format(Locale.ROOT, "%02d", dt.get(IsoFields.WEEK_BASED_YEAR));
+            return yearString.substring(yearString.length() - 2);
+          },
+          "IY"),
+      new NumberFormatPattern(
+          dt -> {
+            final String formattedYear = String.format(Locale.ROOT, "%03d", dt.getYear());
+            if (formattedYear.length() > 3) {
+              return formattedYear.substring(formattedYear.length() - 3);
+            } else {
+              return formattedYear;
+            }
+          },
+          "YYY"),
+      new NumberFormatPattern(
+          dt -> {
+            final String formattedYear = String.format(Locale.ROOT, "%02d", dt.getYear());
+            if (formattedYear.length() > 2) {
+              return formattedYear.substring(formattedYear.length() - 2);
+            } else {
+              return formattedYear;
+            }
+          },
+          "YY"),
+      new NumberFormatPattern(
+          dt -> {
+            final String formattedYear = Integer.toString(dt.getYear());
+            if (formattedYear.length() > 1) {
+              return formattedYear.substring(formattedYear.length() - 1);
+            } else {
+              return formattedYear;
+            }
+          },
+          "Y"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%02d", dt.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR)),
+          "IW"),
+      new NumberFormatPattern(
+          dt -> {
+            final Month month = dt.getMonth();
+            final int dayOfMonth = dt.getDayOfMonth();
+            final int weekNumber = dt.get(IsoFields.WEEK_OF_WEEK_BASED_YEAR);
+
+            if (month == Month.JANUARY && dayOfMonth < 4) {
+              if (weekNumber == 1) {
+                return String.format(Locale.ROOT, "%03d", dt.getDayOfWeek().getValue());
+              }
+            } else if (month == Month.DECEMBER && dayOfMonth >= 29) {
+              if (weekNumber == 1) {
+                return String.format(Locale.ROOT, "%03d", dt.getDayOfWeek().getValue());
+              }
+            }
+
+            return String.format(Locale.ROOT, "%03d",
+                (weekNumber - 1) * 7 + dt.getDayOfWeek().getValue());
+          },
+          "IDDD"),
+      new NumberFormatPattern(
+          dt -> Integer.toString(dt.getDayOfWeek().getValue()),
+          "ID"),
+      new NumberFormatPattern(
+          dt -> {
+            final String yearString = Integer.toString(dt.get(IsoFields.WEEK_BASED_YEAR));
+            return yearString.substring(yearString.length() - 1);
+          },
+          "I"),
+      new StringFormatPattern(
```

Comment:
I am not sure Calcite can handle dates BC, but maybe this code should not concern itself with that point.
But you may not be able to test this codepath end-to-end.